### ISSUE TITLE
C++: extractor fails when compiling with custom clang build

### ISSUE THREAD

**Issue by GJL** [github/codeql#15898]
The C++ extractor (release **2.16.4**) doesn't work anymore in combination with a custom clang build. 

The extractor detects the clang version I am using as 9999.0.0 and adds the arguments `--clang_version 99990000` to its processed command line. Finally it fails with `Command-line error: invalid clang version number: 99990000`.  However, the C++ extractor from release **2.15.0** could handle the special clang version 99990000.

The clang build I am using reports its version in the form of:
```
clang version <company name>-trunk (<SHA1 hash>)
```

Running
```
$ /path/to/custom/clang --version
```
reports something similar to:
```
clang version foobar123-trunk (a94a8fe5ccb19ba61c4c0873d391e987982fbbd3)
```

**Minimal reproducible example (command):**
```
$ /codeql/cpp/tools/linux64/extractor-static --mimic /path/to/custom/clang ~/test.c
```

Output:
```
[E 09:50:35 1652067] Processed command line: /[...]/codeql/cpp/tools/linux64/extractor-static --executable_filename /tmp/test/a.out -w --error_limit 1000 --disable_system_macros --variadic_macros --incognito --gcc --clang_version 99990000 --gnu_version 40801 --has_feature_vector 1111111111111111000000000000000000000000000000000000000000000000000011 --clang --target linux_x86_64 -D_LP64=1 [...] -- /[...]/test.c
Command-line error: invalid clang version number: 99990000

[E 09:50:35 1652067] Warning[extractor-c++]: In construct_text_message: Command-line error: invalid clang version number: 99990000


1 catastrophic error detected in this compilation.
Compilation terminated.
[...]
[E 09:50:35 1652067] Warning[extractor-c++]: In main: Extractor exiting with code 1
```

`test.c` contains a trivial C program:
```C
#include <stdio.h>

int main() {
  printf("hello world\n");
}
```

**Comment by jketema** [comment]
Hi @GJL,

This behaviour changed in CodeQL 2.15.2. The maximum version number that is currently allowed is 99.99.99. The version number is derived from the values of the `__clang_major__`, `__clang_minor__`, and `__clang_patchlevel__` macros, as set by clang. Have you changed those values in your custom build of clang?

**Comment by GJL** [comment]
Hi @jketema,

Thanks for clarifying how the version is determined. Our custom build expands the macros to `9999`, `0`, and `0`, respectively. 

Would it be possible to lift the clang major version limit of 99 introduced in CodeQL 2.15.2 in the next CodeQL release, or are there other workarounds we can employ? 

I think that changing the major version to 99 is risky on our end; the toolchain is used by countless teams and many projects may break.

**Comment by jketema** [comment]
Thanks for the additional info.

> Would it be possible to lift the clang major version limit of 99 introduced in CodeQL 2.15.2 in the next CodeQL release, or are there other workarounds we can employ?

I don't believe there are any workarounds for this. Due to our release cycle, the earliest we can fix this is in CodeQL 2.17.0, which should be the release after next. I'll have to discuss internally though whether we want to do this.

**Comment by jketema** [comment]
We merged a fix for this. It should land in CodeQL 2.17.0, which should be released in about 4 weeks.
